src/Analysis/Ast/Impl/Indexing/IIndexManager.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis.Documents;
+
+namespace Microsoft.Python.Analysis.Indexing {
+    internal interface IIndexManager : IDisposable {
+        Task AddRootDirectoryAsync();
+        void ProcessNewFile(string path, IDocument doc);
+        Task ProcessClosedFileAsync(string path, CancellationToken fileCancellationToken = default);
+        void ReIndexFile(string path, IDocument doc);
+        Task<List<HierarchicalSymbol>> HierarchicalDocumentSymbolsAsync(string path);
```

Comment:
@jakebailey this would be the asynchronous interface for querying the SymbolIndex.
Internally it awaits for the AddRootDirectory Task to complete.

Reply:
Because you've switched these to List, you're still going to need to have limits somewhere in the API. WorkspaceSymbolsAsync simply won't work in a real-world codebase where the LS will attempt to create and serialize a list with every single symbol in a project.

Reply:
It looks like we want enumerable that is dynamically populated asynchronously and the reader of the enumerable should be awaiting on reading next item. Is that the case? How's about BufferBlock<T> ? https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.bufferblock-1?view=netcore-2.2 and its
https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.dataflowblock.receiveasync?view=netcore-2.2#System_Threading_Tasks_Dataflow_DataflowBlock_ReceiveAsync__1_System_Threading_Tasks_Dataflow_ISourceBlock___0__

Reply:
The only reason I used enumerables in the original index was to not expose the limits in the API (and it was less code given that those values were already in `Server.WorkspaceSymbols.cs`), but in an async context it has unclear semantics (due to cancellation and ownership).

BufferBlock is .NET Core only, not .NET Standard, so I don't think we could use it. I had been looking for AsyncEnumerable, but it's [not standardized yet](https://github.com/dotnet/csharplang/blob/master/proposals/async-streams.md).

Reply:
Probably the easiest thing to do would be to have:

```csharp
Task<IReadOnlyList<HierarchicalSymbol>> HierarchicalDocumentSymbolsAsync(string path);
Task<IReadOnlyList<FlatSymbol>> WorkspaceSymbolsAsync(string query, int? maxLength = null);
```

Or similar (pick your default).

The depth limits only used in the `SymbolInformation[]` version of `textDocument/documentSymbol` and really isn't needed.

Reply:
`DataflowBlock` is a code juggernaut that creates lots of problems for debugging. I prefer to stay away from it if I can. `IEnumerable` is dynamically populated by itself, but it will be randomly populated. We can start from opened IDocument instances with available AST's, and then follow up with files that aren't opened yet.

Reply:
We used BufferBlock in RTVS, worked fine. But @AlexanderSher knows better :-)
@AlexanderSher - the problem is with populating enumerable asynchronously up to a limit since VSC does not accept more than ~1000 entries. But yeah, simple limit as @jakebailey suggest will work too.